### DISCUSSION TITLE
Behavior of synchronous path operation functions

### DISCUSSION THREAD

**Discussion by dxnxk-0** [fastapi/fastapi#8990] [Questions] [answered]
Hello,

_Note: I posted this question in "discussions" but not sure it was the right place I apologize for the duplicate_

Coming from Scala where the management of asynchronous/synchronous operations is quite explicit, I am struggling to understand what's the impact of using def instead of async def for path operation functions.

My main motivation by using an async framework is to be able to scale a lot on a single machine, being able to support a high number of concurrent http requests with a limited number of system threads.

Using a mature database access library like SQL Alchemy will require to use non async functions a lot, as it does not support async yet (I know there are options but I am not sure I want to use it). The doc claims that then the function runs on a separate thread pool.

Does that mean:
1/ a new thread will be created for each new http client request: in that case I guess the async nature of fastapi just disappear...

2/ a new thread will be created just for a new sql connection/session while keeping the requests handlers async on the main pool ? In that case what kind of thread is it ? is it a fixed thread pool ? is the number of threads configurable ? How will it be managed regarding the number of connections in the connection pool (dead locks ...) ?

Thanks for your help !
--pascal

**Comment by falkben** [comment, marked as answer]
Sqlalchemy has support for async in beta (1.4). See docs [here](https://docs.sqlalchemy.org/en/14/changelog/migration_14.html#asynchronous-io-support-for-core-and-orm).

1. A new thread is not created for each new request, if you use a sync function, it is run inside a threadpool.
2. The threadpool size is set by the standard library. It is activated by starlette's [`run_in_threadpool`](https://github.com/encode/starlette/blob/e4307065ea6dbe708fba9643d14fe7adfff06d46/starlette/concurrency.py#L27-L40) which calls [`loop.run_in_executor()`](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.run_in_executor) which by default uses [`ThreadPoolExecutor`](https://docs.python.org/3/library/concurrent.futures.html#threadpoolexecutor), which you can see has a default max size, depending on Python version.

I think the [async docs](https://fastapi.tiangolo.com/async/) in FastAPI do a very good job of explaining the practical effects of using async. If you're looking for more information, this [video series](https://youtu.be/Xbl7XjFYsN4) by Łukasz Langa, one of the python release managers, does a great job of going into more depth on it.

**Comment by pikeas** [comment]
+1, I was just about to write up a similar question.

I'm familiar with threading and async, but even after reading https://fastapi.tiangolo.com/async/, it's not clear to me what magic FastAPI is doing under the hood when it comes to making async def/def just work.

The docs say: 

> When you declare a path operation function with normal def instead of async def, it is run in an external threadpool that is then awaited, instead of being called directly (as it would block the server).

FastAPI runs sync functions in a separate threadpool, but the implications aren't at all clear.

Let's say we have:

```
async def async_route(sync_val: SyncVal = Depends(sync_fn), async_val: AsyncVal = Depends(async_fn)): ...

def sync_route(sync_val: SyncVal = Depends(sync_fn), async_val: AsyncVal = Depends(async_fn)): ...

```

1) `async_route` + `sync_fn`: what happens if `async_route` does something that isn't thread-safe? For example, what if `sync_fn` is `def get_db(): session = Session(); yield session`?
2) `async_route` + `sync_fn`: what if `sync_fn` blocks on I/O?
3) `async_route` + `sync_fn`: what if `sync_fn` blocks with a high CPU call? I'd imagine this crushes the whole process, but it may only crush the sync threadpool?
4) `async_route` + `async_fn`: What if `async_fn` accidentally blocks? For example, a session waiting to check out a DB connection from a pool.

Now consider `sync_route` + `async_fn`: On the face of it, this doesn't seem like it should work at all, but it does. At first I thought that the route decorator was async, but both `@app.<method>` and `Depends` are defined synchronously, so it's not at all clear how this runs.

5) `sync_route` + `async_fn`: `sync_route` is in a threadpool, but does `async_fn` run in that same thread from the sync threadpool or is it sent back to the main app thread?
6) `sync_route` + `async_fn`: is this thread safe? What if `async_fn` is the `async def` version of `get_db()`?
7) `sync_route` + `sync_fn`: does this run in the same thread of spin off a new one for each dependency?
8) Other combinations of the above

It's fantastic that FastAPI makes all of this Just Work, but it's bordering on too much magic. I'd almost prefer standard async "flavoring", where introducing an async dependency injection requires an explicit `async def` on the endpoint.

What mental model for performance and failure engineering should we apply when building complex FastAPI applications?

**Comment by chenjr95** [comment]
+1 I also had a similar question (thanks for the thorough write up pikeas!)

@tiangolo or @dmontagu would love to hear your thoughts on the expected behavior for the different combinations of `async/sync_route` + `async/sync_fb`. Thanks in advance!

**Comment by tiangolo** [comment]
Thanks for the help here everyone! 👏  🙇 

Agreed to all what @falkben said. Plus these updates:

Now async stuff is handled by AnyIO. It has an internal dynamic thread pool. It creates threads up to a default max of 40, and it removes them when they are not used for a while.

About the questions:

* Any async function can only be run on the main thread, on the main event loop. There's no other way around in Python. So, if there's an async path operation (route) function or an async dependency function, know that it can only be run on the main thread, on the main event loop, and that's what FastAPI will do.
* If you define a regular `def` function for a path operation (route) or dependency, FastAPI will run it via AnyIO on a thread worker. FastAPI runs dependencies before running the path operation function and then passes the values where needed. If a dependency requires another subdependency, FastAPI runs that first, again, if it's an async function, on the event loop on the main thread (it just `await`s it) and if it's a regular `def` function it runs it on a thread worker.

About "thread safety", there's a problem with that term, it sounds as something very specific, and it was easier to assume it covered everything when the mental model meant that running things concurrently and on threads meant the same thing. But that's no longer the case. We should probably say "concurrent safe" instead of "thread safe".

Also, it is not obvious what it refers to. But the main problem underneath it refers to is that you don't want some concurrent code A to modify the *state* (variables, attributes, values) for concurrent code B. If you use global variables, that's when you have those problems. If one of your functions modifies an attribute in a class (not an instance, but the class), and then something else tries to use that class, you have the same problem, it's like a global variable. If you use local variables that are not shared in any way with other concurrent code, which is what you normally do, then you are safe. If you *read* global variables from multiple concurrent places but never modify that, then you're fine too.

If you use thread local variables (as in Python's [`threading.local()`](https://docs.python.org/3/library/threading.html#thread-local-data)), or libraries that depend on thread-local variables, which were originally made to solve these thread safety issues, you are **not safe**. Thread local variables will keep values for a whole thread, that's it. But concurrent code is no longer attached exclusively and 1-on-1 to threads, async code will all run on the same thread, and if you depend on a thread-local to deal with concurrent pseudo-global variables, all your concurrent code will be using the same pseudo-global variable. And then, if you need to call a regular `def` function (e.g. a `def` FastAPI dependency) in a thread worker, suddenly that will no longer have access to the pseudo-global thread-local variable. You should instead use Python's new [`contextvars`](https://docs.python.org/3/library/contextvars.html), and libraries that depended on thread locals are migrating to contextvars (and they should if they haven't). Contextvars are compatible with async contexts AND threads, so you don't lose anything.

But, you should probably stay away from thread-locals and contextvars and all that stuff as much as possible, use local variables, pass values in parameters, etc. It will make your life soooo much easier, your code easier to debug, and the extra time you spend typing the extra parameter in the function will save you orders of magnitude the same time debugging why a pseudo-global is not working correctly (from personal experience 😅 ).

---

After all that, I do want to improve docs about async, concurrency, threads, contextvars, etc. That will happen at some point. But for now, this long message. 📜 

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.

**Comment by jgentil** [comment]
Is there a way to tell FastAPI to _not_ run a specific non-async dependency function in a threadpool?  It seems a bit overkill to assume that every single dependency that isn't async has potentially blocking IO.  Some dependency functions just instantiate simple values or concatenate strings.

**Reply by methane** [reply]
No way, at least for now.
You can use async function wrapping sync function for dependency.

**Comment by AbhinasRegmi** [comment]
So if we have async_dep for a sync_route will both of them run in main thread ?? or just the async_dep in main thread and sync_route in AnyIO threadpool ??

**Reply by methane** [reply]
> just the async_dep in main thread and sync_route in AnyIO threadpool ??

This.

**Comment by brianmedigate** [comment]
I am using FastAPI in a pretty large async-native codebase, and I am a bit worried about accidentally using `def` instead of `async def` for dependencies which don't do any I/O at all, and creating overhead of running those functions in a thread pool.

We very rarely do anything blocking, and even if we do it is usually one call within a larger dependency or endpoint, so we manually offload it to a thread. So the "automatically run anything sync in a threadpool" functionality is more of a liability than an asset.

Some possible mitigations which would help me:

1. a configurable way to turn off the functionality altogether, and have FastAPI just reject non-coroutine functions
2. a way to explicitly opt-in, per function, with a decorator or kwarg to the path operation decorator or something
3. a configurable way to have FastAPI raise a warning or something when running a sync function in a thread pool if it didn't explicitly opt in

Would something along those lines be helpful to others?

In the meantime I'm monkeypatching and raising warnings in tests.

There are also cases that it is harder to avoid writing as sync dependencies, such as classes as dependencies.
